src/modules/components/address/addressInput/addressInput.tsx

```diff
@@ -0,0 +1,250 @@
+
+    const currentChain = wagmiConfig.chains.find(({ id }) => id === processedChainId);
+    const blockExplorerUrl = `${currentChain?.blockExplorers?.default.url}/address/${value}`;
+
+    const supportEnsNames = currentChain?.contracts?.ensRegistry != null;
+
+    const inputRef = useRef<HTMLTextAreaElement>(null);
+
+    const [debouncedValue, setDebouncedValue] = useDebouncedValue(value, { delay: 300 });
+    const [isFocused, setIsFocused] = useState(false);
+
+    const {
+        data: ensAddress,
+        isFetching: isEnsAddressLoading,
+        queryKey: ensAddressQueryKey,
+    } = useEnsAddress({
+        name: ensUtils.isEnsName(debouncedValue) ? normalize(debouncedValue) : undefined,
+        config: wagmiConfig,
+        chainId,
+        query: { enabled: supportEnsNames && ensUtils.isEnsName(debouncedValue) },
+    });
+
+    const {
+        data: ensName,
+        isFetching: isEnsNameLoading,
+        queryKey: ensNameQueryKey,
+    } = useEnsName({
+        address: debouncedValue as Address,
+        config: wagmiConfig,
+        chainId,
+        query: { enabled: supportEnsNames && addressUtils.isAddress(debouncedValue) },
+    });
+
+    const displayMode = ensUtils.isEnsName(value) ? 'ens' : 'address';
+
+    const isLoading = isEnsAddressLoading || isEnsNameLoading;
+
+    const handleInputChange = (event: ChangeEvent<HTMLTextAreaElement>) => onChange?.(event.target.value);
+
+    const toggleDisplayMode = () => {
+        const newInputValue = displayMode === 'address' ? ensName : ensAddress;
+        onChange?.(newInputValue ?? '');
+
+        // Update the debounced value without waiting for the debounce timeout to avoid delays on displaying the
+        // ENS/Address buttons because of delayed queries
+        setDebouncedValue(newInputValue ?? '');
+    };
+
+    const handlePasteClick = async () => {
+        const text = await clipboardUtils.paste();
+        onChange?.(text);
+    };
+
+    const handleClearClick = () => onChange?.(undefined);
+
+    const handleInputFocus = (event: FocusEvent<HTMLTextAreaElement>) => {
+        setIsFocused(true);
+        onFocus?.(event);
+    };
+
+    const handleInputBlur = (event: FocusEvent<HTMLTextAreaElement>) => {
+        setIsFocused(false);
+        onBlur?.(event);
+    };
+
+    // Trigger onChange property when value is a valid address or ENS
+    useEffect(() => {
+        if (isLoading) {
+            return;
+        }
+
+        if (ensAddress) {
+            // User input is a valid ENS name
+            const normalizedEns = normalize(debouncedValue);
+            onAccept?.({ address: ensAddress, name: normalizedEns });
+        } else if (addressUtils.isAddress(debouncedValue)) {
+            // User input is a valid address with or without a ENS name linked to it
+            const checksumAddress = addressUtils.getChecksum(debouncedValue);
+            onAccept?.({ address: checksumAddress, name: ensName ?? undefined });
```

Comment:
The `ensName` variable here is actually set as `string | null`, the fallback to `undefined` is to avoid emitting a `string | null | undefined` value but just `string | undefined`

Reply:
Ah, the `GetEnsNameReturnType` is `string | null`. I missed that.